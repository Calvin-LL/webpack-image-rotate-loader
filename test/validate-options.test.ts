import webpack from "webpack";

import WIRLWebpackTestCompiler from "./helpers/WIRLWebpackTestCompiler";

describe.each([4, 5] as const)("v%d validate options", (webpackVersion) => {
  const tests = {
    background: {
      success: ["#FFF", "rgb(2,4,99)", { r: 255, g: 255, b: 255 }],
      failure: [false, 9],
    },
    angle: {
      success: [undefined, 90, -90, -190],
      failure: [false, "false"],
    },
  };

  function createTestCase(
    key: string,
    value: any,
    type: "success" | "failure"
  ): void {
    it(`should ${
      type === "success" ? "successfully validate" : "throw an error on"
    } the "${key}" option with ${JSON.stringify(value)} value`, async () => {
      const compiler = new WIRLWebpackTestCompiler({ webpackVersion });

      let stats: webpack.Stats | undefined;

      try {
        stats = (
          await compiler.compile({
            loaderOptions: {
              [key]: value,
            },
            throwOnError: false,
          })
        ).stats;
      } finally {
        if (type === "success") {
          expect(stats!.hasErrors()).toBe(false);
        } else if (type === "failure") {
          const errors = stats!.compilation.errors;

          expect(errors).toHaveLength(1);
          expect(errors[0].error.message).toMatchSnapshot();
        }
      }
    }, 60000);
  }

  for (const [key, values] of Object.entries(tests)) {
    for (const type of Object.keys(values) as ("success" | "failure")[]) {
      for (const value of values[type]) {
        createTestCase(key, value, type);
      }
    }
  }
});
