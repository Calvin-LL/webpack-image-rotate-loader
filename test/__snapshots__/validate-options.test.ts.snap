// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`v4 validate options should throw an error on the "angle" option with "false" value 1`] = `
"Invalid options object. Image Rotate Loader has been initialized using an options object that does not match the API schema.
 - options.angle should be a number.
   -> The angle to rotate the input image. Or \`undefined\` which will use the EXIF \`Orientation\` tag. Default to \`undefined\`"
`;

exports[`v4 validate options should throw an error on the "angle" option with false value 1`] = `
"Invalid options object. Image Rotate Loader has been initialized using an options object that does not match the API schema.
 - options.angle should be a number.
   -> The angle to rotate the input image. Or \`undefined\` which will use the EXIF \`Orientation\` tag. Default to \`undefined\`"
`;

exports[`v4 validate options should throw an error on the "background" option with 9 value 1`] = `
"Invalid options object. Image Rotate Loader has been initialized using an options object that does not match the API schema.
 - options.background should be one of these:
   string | object { … }
   -> The background color of the output image if the image is rotated by an angle other than multiples of 90. Default to \`\\"#000000\\"\`
   Details:
    * options.background should be a string.
    * options.background should be an object:
      object { … }"
`;

exports[`v4 validate options should throw an error on the "background" option with false value 1`] = `
"Invalid options object. Image Rotate Loader has been initialized using an options object that does not match the API schema.
 - options.background should be one of these:
   string | object { … }
   -> The background color of the output image if the image is rotated by an angle other than multiples of 90. Default to \`\\"#000000\\"\`
   Details:
    * options.background should be a string.
    * options.background should be an object:
      object { … }"
`;

exports[`v5 validate options should throw an error on the "angle" option with "false" value 1`] = `
"Invalid options object. Image Rotate Loader has been initialized using an options object that does not match the API schema.
 - options.angle should be a number.
   -> The angle to rotate the input image. Or \`undefined\` which will use the EXIF \`Orientation\` tag. Default to \`undefined\`"
`;

exports[`v5 validate options should throw an error on the "angle" option with false value 1`] = `
"Invalid options object. Image Rotate Loader has been initialized using an options object that does not match the API schema.
 - options.angle should be a number.
   -> The angle to rotate the input image. Or \`undefined\` which will use the EXIF \`Orientation\` tag. Default to \`undefined\`"
`;

exports[`v5 validate options should throw an error on the "background" option with 9 value 1`] = `
"Invalid options object. Image Rotate Loader has been initialized using an options object that does not match the API schema.
 - options.background should be one of these:
   string | object { … }
   -> The background color of the output image if the image is rotated by an angle other than multiples of 90. Default to \`\\"#000000\\"\`
   Details:
    * options.background should be a string.
    * options.background should be an object:
      object { … }"
`;

exports[`v5 validate options should throw an error on the "background" option with false value 1`] = `
"Invalid options object. Image Rotate Loader has been initialized using an options object that does not match the API schema.
 - options.background should be one of these:
   string | object { … }
   -> The background color of the output image if the image is rotated by an angle other than multiples of 90. Default to \`\\"#000000\\"\`
   Details:
    * options.background should be a string.
    * options.background should be an object:
      object { … }"
`;
